### PR TITLE
Add pygments lexer and use it as a custom lexer for chpldoc

### PR DESCRIPTION
**Pull request by e-kayrakli**
This PR copies the Chapel lexer from upstream Pygments and uses it as a custom
lexer. This'll allow us to avoid longer turnaround times with Pygments versions.

Resolves the first bullet in https://github.com/chapel-lang/chapel/issues/14623

Following runs successfully and generates docs as we have it today:
`make -C third-party/chpl-venv chpldoc-venv`
`make docs`

### CODE REVIEW

**Comment by e-kayrakli** [comment]
@ben-albrecht -- As we discussed a little while ago..

I am not sure how to handle the LICENSE, though. We can copy it (as suggested by it) but somehow add a reference to upstream Pygments?

https://github.com/pygments/pygments/blob/master/LICENSE

Looking to see if there is any precedence for this that I can find.

**Comment by ben-albrecht** [comment]
> I am not sure how to handle the LICENSE, though. We can copy it (as suggested by it) but somehow add a reference to upstream Pygments?

I am not sure. I thought there was a formal process for bundling new third-party licensed code into the repo. @bradcray probably knows more about that process.

There are some other files to be updated too such as [`third-party/README`](https://github.com/chapel-lang/chapel/blob/master/third-party/README) and the downstream README.

**Comment by bradcray** [comment]
There is a process we're supposed to go through before importing third-party code into our code base.  In this case, though, did we do all the development on the original Pygments file ourselves?  For the file in its current location, it seems so.  But it looks like it moved from somewhere else at some point, and I didn't trace back to that point.

**Comment by ben-albrecht** [comment]
> There is a process we're supposed to go through before importing third-party code into our code base. In this case, though, did we do all the development on the original Pygments file ourselves?

We developed the file and contributed it to the project under their license (bsd). I don't know if that matters from a legal process standpoint.

**Comment by e-kayrakli** [comment]
So here's what I think I'll do:

- Put the [LICENSE](https://github.com/pygments/pygments/blob/master/LICENSE) as the same location as the lexer. 
- Change the header comment to replace `AUTHORS` with `Thomas van Doren`
- Update `third-party/chpl-venv/README.md` to add the lexer to the list.

Do these make sense?

**Comment by bradcray** [comment]
I think that's reasonable.  The unfortunate thing is that Thomas worked for us at the time and developed this for our usage, so could've just added the file to our repo without the need for a LICENSE file or the like, but I guess that's water under the bridge at this point.

One other potential place to check is the top-level LICENSE file which summarizes other third-party licenses used within the project.  There's already a Pygments entry, so we might be good, but double-check me.

**Comment by e-kayrakli** [comment]
You're right. We have an entry for Pygments there and refer to READMEs in relevant subdirs. I think, that, and what I proposed above complements one another (with the disclaimer that I don't know much about legal details of such matters). To say the least, I am not sure what else we could do.

I'll update this PR to make those changes.

**Comment by e-kayrakli** [comment]
@ben-albrecht -- I have updated the PR with some changes for licensing. You have said 

> There are some other files to be updated too such as third-party/README and the downstream README.

Our third-party/chpl-venv/README.md already refers to Pygments but should we add a separate header for this lexer? As for the downstream README, I am not sure what you mean exactly.

**Review by ben-albrecht — APPROVED** [review]

**Comment by ben-albrecht** [comment]
@e-kayrakli - Approved, but is the CI failure expected?

**Comment by ben-albrecht** [comment]
> Our third-party/chpl-venv/README.md already refers to Pygments but should we add a separate header for this lexer? As for the downstream README, I am not sure what you mean exactly.
 
I think I was assuming there'd be a `third-party/chpl-venv/sphinxcontrib-chapeldomain/README` added as well, but (1) that may not exist upstream, and (2) may not be necessary anyway.

**Comment by ben-albrecht** [comment]
> Our third-party/chpl-venv/README.md already refers to Pygments but should we add a separate header for this lexer? 

I don't think so.

**Comment by bradcray** [comment]
> I think I was assuming there'd be a third-party/chpl-venv/sphinxcontrib-chapeldomain/README added as well

@ben-albrecht: Are you imagining that if we wrote this, it might capture the provenance of the code using a description like "This is a local snapshot of a Pygments highlighter that Thomas Van Doren originally developed as a member of the Chapel team and contributed to the Pygments project at URL.  In November 2020, we decided to bring a copy of it back into our source tree in order to blah blah blah."?  I could imagine that being useful, if so, and would get behind it.

**Comment by ben-albrecht** [comment]
> Are you imagining that if we wrote this, it might capture the provenance of the code using a description like "This is a local snapshot of a Pygments highlighter that Thomas Van Doren originally developed as a member of the Chapel team and contributed to the Pygments project at URL. In November 2020, we decided to bring a copy of it back into our source tree in order to blah blah blah."? I could imagine that being useful, if so, and would get behind it.

I was more imagining that a README exists in the upstream repository and we copy that over as well. However, I like your idea better.

**Comment by e-kayrakli** [comment]
>  but is the CI failure expected?

It is not, and I think (but not sure) it may have something do with venv changes as I was able to run make docs for sure before.

>Are you imagining that if we wrote this, it might capture the provenance of the code using a description like "This is a local snapshot of a Pygments highlighter that Thomas Van Doren originally developed as a member of the Chapel team and contributed to the Pygments project at URL. In November 2020, we decided to bring a copy of it back into our source tree in order to blah blah blah."? I could imagine that being useful, if so, and would get behind it.

Will add this.

**Comment by e-kayrakli** [comment]
Alright, there has been some major changes to make packaging more reasonable.

- Now we have a `chpllexer` local package that we install during `make chpldoc-venv`
- This package has only one file and that is identical to what is currently contained in Pygments. (Subsequent PR https://github.com/chapel-lang/chapel/pull/16562 will make some changes to this, but it needs to be adjusted)
- The package's README has the explanation above. And LICENSE file that is the BSD license that Pygments uses.

I think this makes packaging look more reasonable, what do you think?

One lingering question is: how should we version the lexer? Grab the sphinxcontrib version? Somehow version it separately? I don't have a strong opinion but the latter may be more principled. It may also help track the differences between downstream (Pygments) and upstream (local) lexers.

**Comment by e-kayrakli** [comment]
I should probably add some links to some resources that I used. But that's pending if we want to move forward with this approach

**Comment by ben-albrecht** [comment]
> I think this makes packaging look more reasonable, what do you think?

I agree. I like this form much better.

re: versioning -- where is the version specified and/or used? Does the chapel lexer have a separate version from Pygments itself?

If I understand [PEP 440](https://www.python.org/dev/peps/pep-0440/#local-version-identifiers) correctly, we can use `x.y.z+[0-9]*` to denote revisions from an upstream release, starting with `+0` if no revisions have been made.

**Comment by e-kayrakli** [comment]
Ah that's good. The version is currently not specified so when installed it looks like `chpllexer-0.0.0`, and I am not sure if there is any practical use for it beyond making things a bit easier when synching with the pygments repo.

I'll version this as `2.7.2+0` as it is forked from pygments 2.7.2, subsequent improvements we have locally will be 2.7.2+1 etc etc.

(Just for the record, https://www.python.org/dev/peps/pep-0440/#adding-local-version-identifiers is the part where that PEP suggests that)

**Comment by e-kayrakli** [comment]
Alright, with the latest set of commits, I think this is ready to go in.

@ben-albrecht -- thanks for the version suggestion, this currently sets the lexer version to `2.7.2+0`

I'll kick few more tires tomorrow, update the PR message and merge, unless there's any concern or issue.

File: third-party/chpl-venv/chpllexer/README.rst [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+In November 2020, we decided to bring a copy of it back into our source tree in
+order to avoid lags in published pygments versions and shorten the time between
+any improvements to this lexer and seeing their impact on the generated
+documentation.
+
+See https://github.com/chapel-lang/chapel/issues/14623 for motivation.
+
+Changes made to chpllexer/chapel.py should be reflected downstream for making
+the improvements publicly available through Pygments.
```
Comment (ben-albrecht) [inline third-party/chpl-venv/chpllexer/README.rst:13]: Might want to make a note explaining the version numbers in here

Comment (e-kayrakli) [reply third-party/chpl-venv/chpllexer/README.rst:13]: Good point

**Comment by e-kayrakli** [comment]
Updated. I think this is good to go, but I'll wait until other failures regarding python versions resolve.

**Comment by e-kayrakli** [comment]
I am running into issues with this after the recent changes:

- sphinxcontrib is not bundled anymore, so we can't have a local version diverging to pick the custom lexer
- pygments allows installing lexers as plugins and even allows them to override builtin lexers through different means. However, this is only implemented for automatically inferring the lexer from the filename. Whereas, sphinx says "go get me `chapel`" , and for that pygments just returns the first `chapel` lexer it can find, and that is always the builtin one. I opened an issue about this on pygments repo. But don't know how quickly they respond. https://github.com/pygments/pygments/issues/1603

Unfortunately, we'll need to rethink how to incorporate this again :( I think the most reasonable solution is to modify sphinxcontrib to check if `chpllexer` package is installed and if so, use the lexer from it explicitly (this'll override the builtin pygments lexer). For those who just pick our sphinxcontrib, they'll still be using the pygments' lexer. Whereas, we'll install our own "custom" lexer with venv, and sphinx would use that.

The only drawback of that is that we'll have to update sphinxcontrib and bump its version in the venv requirements. But I don't think that's an issue.

@ben-albrecht @lydia-duncan thoughts?

**Comment by lydia-duncan** [comment]
> I think the most reasonable solution is to modify sphinxcontrib to check if chpllexer package is installed and if so, use the lexer from it explicitly (this'll override the builtin pygments lexer). For those who just pick our sphinxcontrib, they'll still be using the pygments' lexer. Whereas, we'll install our own "custom" lexer with venv, and sphinx would use that.
>
> The only drawback of that is that we'll have to update sphinxcontrib and bump its version in the venv requirements. But I don't think that's an issue.

This seems like a reasonable plan to me, and I am also not concerned about having to update that repo and bump its version number, that was not too bad when I did it recently (though we may require Jenna, Soohoon or me to do it, since you need administrator privileges to update PyPI)

**Comment by lydia-duncan** [comment]
Sorry this has been such a pain for you :(

**Comment by ben-albrecht** [comment]
@lydia-duncan - offline, @e-kayrakli and I discussed going back to the solution of bundling the Chapel lexer into our sphinxcontrib repo (as a dependency), which feels a lot cleaner than the direction above. Do you have objections to that?

**Comment by e-kayrakli** [comment]
With the recent changes, this PR is moot. A mirror version of this is added to sphinx's chapel domain via https://github.com/chapel-lang/sphinxcontrib-chapeldomain/pull/40

**Pull request closed without merging**